### ISSUE TITLE
Help requested to build URLs to display simple GVE diagrams for single persons in a story also containing URLs for saved settings

### ISSUE THREAD

**Issue by schuco** [Neriderc/GVExport#347]
I have prepared the following content of a simple menu:

![grafik](https://user-images.githubusercontent.com/5896925/217630347-eb916f71-bdbe-4289-a9ef-9b5758d778a3.png)
--------------------------------------------------------------------------------------------------------------------------------------
Blue text icontains links.
The first 3 lines in blue contain URLs to display a simple diagram for one person.
They look like: 
https://brandneu.schulte-coerne.de/index.php?route=%2Fmodule%2F_GVExport_%2FChart%2Ftree2&xref=I1311
I deliberately did not use saved settings here, because when you have many of such links it would be too much effort to prepare these and overload the list of saved settings. The "_simple links_" can be inserted easily with cut and paste. For different persons only the XREF has to be adapted.

The last blue line (Familiengeschichte Dieckhöfer) includes an URL with saved settings having lots of starting persons.
The second blue line calls a story (webtree feature) where the page of the story containes _simple links_ as described and URLs with saved settings.

My problem:
after I have clicked the last blue line with saved settings displaying many starting persons and thereafter click one of my "simple links" the previous starting persons remain together with the person intended to be displayed.

**Is there a way to build simple links which can be easily handled (cut and paste, just adapting XREF)  where multiple starting persons in a preceeding diagram do not disturb?**

**Comment by Neriderc** [comment]
This is related to previous discussions on how to handle the starting persons list when coming from other areas of webtrees. If you are on a page of Joe Bloggs with XREF "X1", if you click Charts->GVEXport in order to go to GVE, webtrees will set in the URL xref=X1 so we can load that person.

Currently, if there was just one person in the starting list, they are replaced with the XREF from the URL. If there are multiple, they are added to the list. This was so you could incrementally add to your list, and because webtrees always provides an XREF and so you could never save a list of starting persons if they were overwritten by the XREF in the URL.

So the behaviour you see is expected. But that doesn't help you solve your issue :slightly_smiling_face: 

It would be simple enough to add a parameter to the URL that indicates this XREF should override any saved persons. I can not think of a nice way to put this in the UI, but at least initially we can have it as a hidden feature that you need to add to the URL yourself and if we think of a good way to add it to the UI in future we can do that.

How does that sound? For example, we could call it "overwrite", and so you could add "&overwrite=1" to the end of the URL and this would be a sign to GVE to overwrite the list of persons instead of adding to it. I say "=1" because this is how the reset parameter works, if the URL has reset=1 then the settings are reset.

How does that sound?

**Comment by Neriderc** [comment]
That's a great idea! We had always intended to revisit this after using the current setup for a while, to see if it would still be necessary. It seems we now have a good argument to add the option. And I prefer this over a URL parameter which most people would not know exists.

I think adding this to the advanced section of "People to be included" makes the most sense. However, I don't have any ideas on what to call this option.

If you come from any other page in webtrees, the relevant XREF will be fed to GVE. E.g.  If you come from an individual's page, their XREF is fed in to GVE. If you come from most other pages, the default individual for the tree is used.

What would the name for this option be?

"What should happen with the individual that is fed to GVE from webtrees when loading the GVE page?" is a bit wordy.

It could be a dropdown list, default to what we do now (which could just be called "Default", but also add options "Add to list", "Don't add to list", "Overwrite list".

**Comment by hartenthaler** [comment]
"Management of source of individuals?" could be a name for this setting.

**Comment by schuco** [comment]
In our discussions here we mostly and correctly speak about _"starting persons"_ and less about _"people to be included"_, when we adress the individuals in the list _"include anyone related to"_.
@hartenthaler 's idea _"management of source individuals"_ comes close to _"management of starting persons"_. I would like both.
But we should introduce and use the term _"source individuals"_ resp. _"starting persons"_ also in other places of the menu.  So it could be added in brackets:  _"include anyone related to **(source individuals/starting persons)"**_. 
Then we could write _"XREFs of source individuals/starting persons"_ instead of _"XREFs of included individuals"_, which would be more correct anyway.

**Comment by hartenthaler** [comment]
I think the wording of the drop down list values is already very good. We are searching for the related question. 

My suggestions as first ideas (but my English is not good enough for that):

What happens with new and existing start individuals?

Are new start individuals replacing existing ones?

**Comment by Neriderc** [comment]
I think using both "source individuals" and "starting persons" takes up a lot of precious space that we want to use carefully.

If "source individuals" is a more correct term, I suggest we just change references to "starting persons" to instead be "source individuals".

Now as for naming the setting, how do we make it clear that the setting is for the management of the XREFs that come via the URL, and doesn't affect the individuals chosen using the selection box?

**Comment by schuco** [comment]
Pobably: 
_How to set source individuals?_

**Comment by hartenthaler** [comment]
"source individuals by call"
"transmitted source individuals"
"via URL transferred source individuals"

**Comment by Neriderc** [comment]
Sorry just to be clear, is the term "source individuals" supposed to refer to any starting individual, or was the thinking that only the URL-supplied XREF was a "source individual"? (the page you come from is the "source", and the XREF provided by that page is the "source individual"?)

**Comment by hartenthaler** [comment]
Difficult. I would say "source individuals = any starting individual" and I would use "existing individuals" versus "new individuals".

1. Start person / Starting person / Source person: for me all three versions are ok. These persons build a list of persons as a base for all then added persons in the diagram (partners/siblings/ancestors/...).
2. Stop person / Stopping person: When persons are added to the diagram and such a person is reached the process stops at that point. The person is added, but no partners/siblings/ancestors/...

So maybe start/stop is a better pair for the wording than source/stop.

Now the new point. Let me describe how it is implemented now (following my understanding in order to make it clear to me). There are four ways to add start/source persons to the list:

a) navigate to the individual page of a person and open GVExport; the XREF of this person is transferred via the URL to GVExport.

- if there is no or only one existing start/source person in  GVExport: the new person from the individual page is now the only one
- if there are already more than one existing start/source person in GVExport: the new person from the individual page is added to the list of the existing ones in GVExport

b) open GVExport when you are not on a individual page; there is no XREF transferred to GVExport.

- if there is a default person for this tree, this one is used and case a) is used for the further processing
- if there is no default person defined for this tree, the existing start/source person list in GVExport is not touched

c) if you call a saved setting of GVExport (via URL or open it in the user interface of GVExport), then the eventually stored XREFs of start/source persons in the setting are used to replace or to be added to the existing start/source persons following process a).

d) if GVExport is already open you can add and delete start/source persons manually using the functions in the user interface of GVExport

Now we are discussing a new option (in the advanced section of "People to be included"):

> "What should happen with the individual that is fed to GVE from webtrees when loading the GVE page?" is a bit wordy.
> It could be a dropdown list, default to what we do now (which could just be called "Default"), but also add options "Add to list", "Don't add to list", "Overwrite list".

In my opinion, we need a new option for the before-mentioned cases a), b) and c):
"**What should happen with new individuals that are fed to GVExport?**"

Depending on the four values of the new option; in case c) the value of the new option in the setting is used, not the actual value of the option:
1. Option "Default": all cases as described above

2. Option "Add to list":

- case a) the XREF transferred via URL is added to the existing list
- case b) if a default person is defined for this tree this person is added to the existing list
- case c) the XREFs in the setting are added to the existing list

3. Option "Don't add to list": in all cases the existing individuals are not modified; as long as the option is set to this value there is only case d) available to modify the list of start/source persons (or loading a setting in that the value of the new option is not "Don't add to list").

4. Option "Overwrite": the existing list is cleared and then the steps as for option "Add to list" are processed.

**Comment by Neriderc** [comment]
I believe that if there is no default individual for the tree, webtrees still picks one. For GVE2.1.16 and earlier, GVE will not function without an XREF in the URL. My main dev tree has no default individual, but webtrees still passes in XREF X1, which I guess is the root of the tree. So because of this, there isn't an option to not touch the list if there is no default individual. 

But for the rest I believe you have it correct. 

I am not sure we will ever find a good name for the setting. I suggest we just pick one, and use the help text to explain what it means.

What about we go with "Treatment of source individuals"? Based on your first reply but I've put "Treatment" instead of "Management", because it seemed better to me for some reason that I have trouble explaining. And it helps that it's a shorter work.

If there are no objections, we'll go with this, explain it in the help text, and it will be in the advanced section so should confuse and new users.

**Comment by hartenthaler** [comment]
When translating "Treatment" into German this is  better than "Management", too. Sounds good.

**Comment by hartenthaler** [comment]
I just tested the behaviour of webtrees:
- if you call a diagram from the individual page: the XREF of that individual is transferred to the diagram
- if you call a diagram from any other page
  - if there is a default person defined for that tree, the XREF of the default person is transferred to the diagram
  - if there is no default person defined for that tree
    - if there is any person in that tree, the first person in the GEDCOM is transferred to the diagram
    - otherwise (tree is empty) there is an error message ("person does not exist or you do not have the rights")

So you are right.

**Comment by Neriderc** [comment]
I've implemented this now, hopefully as expected.

@schuco let me know if it doesn't solve your issue. In thinking about this, it's possible it doesn't 100% solve your issue as this setting isn't set in the "simple" links, so others may not have this setting set and so

**Comment by schuco** [comment]
Great. It works fine.

Just to the wording:
We have a section: _People to be included_
People to be included are all individuals displayed in the diagramm

some lines belows we write: _XREFs of included individuals_
This is not quite correct. It should be _XREFs of the source individuals_

**Comment by schuco** [comment]
You are right my issue is not completely solved. My concern was  and is that a non experienced user may not be aware which way the source individuals are added. Would  there be a way to just display the present treatment in the normal settings?

**Comment by hartenthaler** [comment]
@schuco: I don't understand your concerns. 

> Rather it is when you click an indi and return to GVE from the indi page.

When I'm on the individual page of a person (for example I123) and call GVExport I expect that something happens: the GVExport diagram opens. ok.
Depending on the new option:
- "Default" and there was no or only one source individual before: I123 is the only source individual. That is what I expected. ok.
- "Default" and there were already two or more source individuals before: I123 is added to the list. Maybe I'm a bit surprised to see still the old source individuals, but I see the old ones and I123 all very prominent and if I don't like the old ones, I can delete them easily. ok.
- "Add to list": see the topic above. ok.
- "Don't add to list": then I123 is not set as source individual. That would be strange. But a "normal" user should never set the new option to "Don't add to list".
- "Overwrite": this is like other diagrams are working, so no surprise. ok.

**Comment by hartenthaler** [comment]
I defined a case c) above:

> c) if you call a saved setting of GVExport (via URL or open it in the user interface of GVExport), then the eventually stored XREFs of start/source persons in the setting are used to replace or to be added to the existing start/source persons following process a).

But this is now implemented in a different way, which is ok for me. But to be precise: if a setting is loaded, then independent of the value of the option "Treatment of source individuals" the new source individuals from the setting are always overwriting the existing list of source individuals.

**Comment by schuco** [comment]
@hartenthaler my concern is not that GVE would not work as it should. My concern is that an unexperienced user who is not aware of the implemented default behaviour may get confused. If such user has e.g. built a GVE tree with several source indis and then selects an individual page to start a new GVE tree from the scratch, he will be surprised to see a mixture of source indis, the one he expected and others he did not expect. Will each user understand GVE's behaviour intuitively just by doing, i.e. learning that two source indis in the preceding setting will result in accumulation whereas having had just one source indi results in overwriting? In my opinion the default behaviour is well chosen, but not obvious. 
Unfortunately I have no good idea how to solve my problem. May be we can add  a status message like "further souce individuals will be added" or "further source individuals will overwrite existing ones" in the bottom bar besides "download" "Reset" and "Help". However my proposed wording is too long and not even precise.

**Comment by hartenthaler** [comment]
In my opinion the result is always obvious, even when a user expects something else. 

But maybe there should be a pop-up status message to indicate what's going on. Like the existing messages "5 persons and 3 families created" there can be a message "one new source person added to 2 existing source persons".

**Comment by Neriderc** [comment]
Ok so to close this issue, all that is left is to:
- Add new option to admin page
- Add message when XREF is loaded from URL

Is that all?
